driver/labjack/reader_source.cpp

```diff
@@ -0,0 +1,480 @@
+        if(err != freighter::NIL) {
+            LOG(ERROR) << "[labjack.reader] Error retrieving channel: " << err.message();
+            continue;
+        }
+        channel.data_type = channel_info.data_type;
+        this->reader_config.index_keys.insert(channel_info.index);
+    }
+
+    for (auto &index_key : this->reader_config.index_keys)
+        keys.push_back(index_key);
+    return keys;
+}
+
+
+void labjack::ReaderSource::init(){
+    if(this->reader_config.device_type != ""){
+        this->init_stream();
+        this->init_tcs();
+        return;
+    }
+    auto [dev, err] = this->ctx->client->hardware.retrieveDevice(
+            this->reader_config.device_key
+    );
+    if(err != freighter::NIL){
+        LOG(ERROR) << "[labjack.reader] Error retrieving device: " << err.message();
+        return;
+    }
+
+    if(dev.model == "LJM_dtT4"){
+        this->reader_config.device_type = "T4";
+    } else if(dev.model == "LJM_dtT7"){
+        this->reader_config.device_type = "T7";
+    } else if(dev.model == "LJM_dtT8"){
+        this->reader_config.device_type = "T8";
+    } else {
+        LOG(ERROR) << "[labjack.reader] Unsupported device type: " << dev.model;
+        return;
+    }
+    this->init_stream();
+    this->init_tcs();
+}
+
+void labjack::ReaderSource::init_tcs(){
+    if(this->reader_config.tc_channels.empty()){
+        return;
+    }
+    {
+        std::lock_guard<std::mutex> lock(labjack::device_mutex);
+        if(check_err(LJM_Open(LJM_dtANY, LJM_ctANY, this->reader_config.serial_number.c_str(), &this->handle), "init_tcs.LJM_OPEN")){
+            LOG(ERROR) << "[labjack.reader] LJM_Open error";
+            return;
+        }
+    }
+
+    for(auto &channel : this->reader_config.channels){
+        if(channel.channel_type ==  "AI"){
+            this->check_err(
+                    LJM_eWriteName(
+                            this->handle,
+                            (channel.location + "_RESOLUTION_INDEX").c_str(),
+                            0
+                    ), "init_tcs.LJM_eWriteName.RESOLUTION_INDEX"
+            );
+            if(this->reader_config.device_type == "T7" || this->reader_config.device_type == "T8") {
+                this->check_err(
+                        LJM_eWriteName(
+                                this->handle,
+                                (channel.location + "_RANGE").c_str(),
+                                0
+                        ), "init_tcs.LJM_eWriteName.RANGE"
+                );
+            }
+            if(this->reader_config.device_type == "T7") {
+                this->check_err(
+                        LJM_eWriteName(
+                                this->handle,
+                                (channel.location + "_NEGATIVE_CH").c_str(),
+                                channel.neg_chan
+                        ), "init_tcs.LJM_eWriteName.NEGATIVE_CH"
+                );
+            }
+        }
+    }
+
+    int msDelay = 1000;
```

Comment:
Not clear to me why this exists. Can you add a comment?

Reply:
resolved